### ISSUE TITLE
Implment KeePass integration

### ISSUE THREAD

**Issue by kmscode** [mRemoteNG/mRemoteNG#187]
https://mremoteng.atlassian.net/browse/MR-490
Add the possibility to link KeePass entries to mRemoteNG and connect the entries to different servers.

It eases up the administration when changing passwords in a multi-domain environment and adds some extra security.

Remote Desktop Manager has this feature and it is pretty neat. 

https://mremoteng.atlassian.net/browse/MR-200
I know some other programs (like Keepass) emulates this feature with a macro which press the keyboard keys automagically to bypass the lack of control with putty.
Example : customize the prompt. We are a bunch of people who access only one account on the server. I can't customize my prompt, alias, etc. without interfere with other people.

**Comment by adam-smith-sundance** [comment]
PAM for linux also has this feature.  It is a similar remote management system but is based on perl

**Comment by trentasis** [comment]
It is working thisfeatures?
It will be great to have this feature

**Comment by kamahat** [comment]
would be nice

**Comment by jotatsu** [comment]
Long term candidate, support for external password keepers

**Comment by MisterY** [comment]
[Kee](https://github.com/kee-org/browser-addon) Firefox add-on uses [KeePassRPC](https://github.com/kee-org/keepassrpc/releases/) to communicate with KeePass and generally works quite well.
How much of an improvement over KeePassHttp this is I'm not sure but both are examples of communication with a running instance of KeePass application.

**Comment by mjmooney** [comment]
This really is the single biggest missing piece for me, as I use a heavily tweaked version of both KeePass and mRemoteNG, and while I can get by with KeeAgent for SSH keys, I simply haven't been able to find any good go-between for credentials, and resort to saving or using modified auto-typing out of KP.  I can also vouch for KeePassRPC and have had very few issues in general with similar KP plugins.

My use case.....  I've integrated 90% of my daily admin environment and toolset inside of mRemote; I use it to integrate everything from web apps and external utilities, as launcher for a library of powershell, bash and other scripts, as well as a wrapper for notepad++ and VSCode + more. Sometimes the performance is clunky or slower (especially when it involves IE or Gecko), but it works most of the time, and it's made work much more streamlined.  This integration would bring the last major piece into one app for me (and many others), and would bring the added security bona-fides of KeePass to mRemoteng.

**Comment by RastaChaum** [comment]
Please ! mRemoteNG + Keepass would be so nice !

**Comment by RFlipper** [comment]
I would be happy to support by money someone who will start the implementation of this feature

**Comment by locke45** [comment]
This Feature would be so nice for our Company Use Case...

**Comment by hummigbird1** [comment]
I am considering to try and implement this feature. I am not promising anything yet, though ;)
Maybe anyone who wants to have this feature can describe (from a users perspective) on how he expects KeePass integration to look and feel and behave?

**Comment by jul-fls** [comment]
Hello, i would like to have this feature on mremoteng, does anyone now if this had been achieved or not ?

**Comment by 4SiB** [comment]
> Hello, i would like to have this feature on mremoteng, does anyone now if this had been achieved or not ?

? At this post you have info on top-right > Projects... that is in TODO at v1.78, check this:
https://github.com/mRemoteNG/mRemoteNG/projects/4#card-38823617

**Comment by jul-fls** [comment]
Ah OK sorry for that...
And do you know when the v1. 78 will be out ?
Best regards.

**Comment by brad457** [comment]
I would be interested in building out this feature. But as @hummigbird1 [asked back in 2020](https://github.com/mRemoteNG/mRemoteNG/issues/187#issuecomment-622948522) if someone could provide a brief description (from a users perspective) how they see mRemoteNG integrating with KeePass best.

We need to have a clear understanding of what is being requested here.

I have used mRemoteNG for long periods of time over the last 7 years and various versions of KeePass over the past 5 years( Windows, Android, IOS, and Linux).  I am familiar with how they work and currently use both daily.

I read through all issues that were linked and marked duplicate with this ([MR-497](https://mremoteng.atlassian.net/browse/MR-497), [MR-490](https://mremoteng.atlassian.net/browse/MR-490), [MR-712](https://mremoteng.atlassian.net/browse/MR-712), [MR-184](https://mremoteng.atlassian.net/browse/MR-184), [MR-900](https://mremoteng.atlassian.net/browse/MR-900)) to try and get a sharper focus on how this should work.

All I know for sure is that users want to be able to have the option added to use a KeePass DB somehow with mRemoteNG.

Are we talking about 
1.  Having the connections in confCons.xml file for mRemoteNG stored in a KeePass DB.
2.  Storing only the credentials that mRemoteNG uses for connections in a KeePass DB.
3.  Something else.

Personally I like the idea of doing ⋕2 as a stepping stone to ⋕1.  Since you can add lots of additional fields and embed files in KeePass, I believe it would be possible to store all the fields associated with an mRemoteNG Connection object.

I would be interested in doing this in the two steps I suggested if that is what the rest of the mRemoteNG community wants.  Thoughts anyone?

**Comment by nemec** [comment]
@brad457 I can't speak for anyone else, but point 2 is definitely what I want to see to consider this request "completed". Point 1 would be nice to have, but not a requirement. The way I imagine it, this could be implemented in roughly one of two ways:

1. Opening a "KeePass integrated connection" prompts for the user's KeePass master password, then mRemoteNG unlocks the KeePass database, pulls the credentials for the entry, and then connects with them. This option requires users to trust mRemoteNG with their master password, which may not be the best option, but IMO it's easiest to implement. Since the original KeePass is a C# app, you can reference it from another .NET application and programmatically open/read KeePass database entries with it.
2. Second option is to somehow integrate with an open (and unlocked) KeePass instance via some sort of inter-process communication. I think [keepasshttp](https://github.com/pfn/keepasshttp) is a common plugin used for this, but it may require users to store the target hostname in the database, too.

**Comment by 4SiB** [comment]
I not want to give mRemoteNG password to unlock database of KeePassXC.
I want register KeePassXC like/similar to a browser plugin, in the same metodology. Then mRemoteNG is like trusted app inside KeePassXC.
Next in login/password fields of mRemoteNG host, best way if this will be searched in KeePassXC database per %HOSTNAME% and grab/pass username/password from it.

Example Username/Password in KeePassXC of some customer's servers, each host have differ password from generator:
![image](https://user-images.githubusercontent.com/18070741/133349473-e14cc107-e69e-4962-96ca-40fcdbbaad25.png)

Now I must leave empty password in mRemoteNG connections, popup show and here I must put password, by manual search proper host from my host list in KeePassXC.

I hope this show, How I hope this will be works in future.

-------------

One mRemoteNG confCons.xml file I store on shared storage, who we share with second person the same base of connections.
I and him we use own username and password from ADDC domain and we cannot put that data info a folder full of connections and enable inheritance in connections, because if we do that then  I login on him account, and opposite :(. Now we cannot put data from system variables or other place and a .conf file store one password, I cannot switch a personality of profile.

If this a multi user works will get from a KeePassXC, then I can have own KeePassXC with my own credentials, and this will solve our problem.
This is like integration with KeePassXC for put details at folder for inheritance connections used this data.

I hope both scenario is common with other users too.
BR Marcin (SiB) Przysowa

**Comment by brad457** [comment]
I appreciate all the input.  I was writing a reply to each of the proposed integrations on my phone but Chrome refreshed. I am going to install KeePassXC and the most popular ports of KeePass over the next few days and do some digging.

Then I will then submit a viable way(s) of carrying this functionality out.

**Comment by Kvarkas** [comment]
From my point the flow may be implemented in such way (I am not familiar with keepass at all, but just follow recommended best practice) but short explanation that I plan for mRemoteNG first.  
It will allow save connections in 3 ways. (but to make that happened need implement profiles first) 
a) Connection data will be stored locally and complete (no questions it will be encrypted for sure) that include all settings for instance and user name, password, port, key... etc. all that is need to complete a connection.  
b) Connection data will be stored partially, so locally we store only host name, IP, settings but no user name, no password, no port or key... to get them user will have 2 ways - enter them manually every time then he make a connection, import them from 3rd party source. That can be DBs or another source. 
c) Connection data will be stored on 3rd party completely
For b and c cases I would like to have ability to separate access to readOnly and readWrite mode. So if user trust app (I mean in term of security terms) it may provide write access to mRemoteNG so it will be able to create DB structure, save data, edit data etc. or that can be provided one time only (like root access) but will not be saved and used only initially then no any other connection data is stored and mostly to prepare structure in DB, and it may be read access then structure or/and values of connection are controlled externally and mRemoteNG only read them then need to make a connection (also I think it will be setting to set expiration data, so let say after specific date write access will be disabled and saved data will be removed from mRemoteNG db. 
Hope you not yet tired to read such :) so now about exact topic - regarding Keepass integration - c case not applicable as Keepas only store user/password/key values, so we have to work on case b. I am not have been look to KeePass api but believe it should be possible to get authentication token from it, so again we have to play scenarios: b1 - then user trust mRemoteNG and save master password with it. On establish connection mRemoteNG will check if he have all data in, as user name and password will be not set but data provider will be set KeePass it will use stored master password to get access to KeePass (getting token, or another way of authorisation that Keepass support) and by that will find required host and load data in (remember in that case data will be used only to establish connection and not be saved even temporary). If we have b2 case then user dont trust (or again security policy not allow to save master passwords in 3rd party app that in such case will be mRemoteNG) then we show up a form to request such data and use it to be authorised without saving it anywhere. And after we establish connection all data not set to be saved in mRemote will be dismissed. 
So long story short. We need find out that authorisations ways support by Keepass, that api calls it allow to make it and then implementing logic will be not very complicated. Sorry for long read :D

**Comment by JoaoPedroAlex** [comment]
I see here a lot of comments , but no instruction on how to integrate mRemoteNG with keypass to be able to authenticate using a single security DB.

**Comment by theMagicRabbit** [comment]
I'm not seeing a consensus one which version of KeePass to support. From what I can see, there appear to be 4 main versions of KeePass. The original KeePass and KeePass 2 from Dominik Reichl (both of which appear to be actively maintained.) Then there is the KeePassX fork, which was again forked into KeePassXC. XC is the only one I see explicitly mentioned by name in this topic. 

XC doesn't have an API or a pluggin infrastructure and according to [their FAQ](https://keepassxc.org/docs/#faq-general-plugins), they have no intention in adding one. They do have a browser extension (for chromium based and firefox) and they suggest using the browser extension's API for native apps (mRemoteNG) that need to communicate with the credential db.

[KeePassX is an abandoned project as of Dec. 2021.](https://www.keepassx.org/index.html%3Fp=636.html)

KeePass 1 & 2 both have a plugin infrastructure. Additionally, KeePass 2 has a scripting plugin that maybe could be used, but it seems like a clunky method to me.

Another possibility (that sounds like a bad plan to me, but I want to mention for the sake of completeness) is to add native support for .kdbx files to mRemoteNG.

I wonder if it might not be easier to work on integrating cloud password managers first and then local password managers like KeePass second. 1password has a developer API for this kind of thing and I assume that other popular cloud based systems have similar APIs as well.

**Comment by support-he** [comment]
I think, if you implement KeePass2, this will be enough, its the Original, and the kdbx-files are the same as the containers from KeepassXC

**Comment by AN1MATEK** [comment]
Just to add my 2 cents: KeePass 2 is the "hardcore" tool for sysadmins. XC is a fork from X that looks good but misses a lot of functionality and it was originally planned for MacOS iirc. KeePass 2 is robust, actively maintained, has tons of options, plugins, etc. If you see people speaking about just KeePass, this means the version from Dominik.
I really hope this will be implemented some day. Thanks to all the developers for the excellent work.

**Comment by AN1MATEK** [comment]
Well I come back because I have had a dumb idea that will integrate KeePass with mRemoteNG "easily". 

KeePass allows you to run commands from the URL field. So for example my URL is ssh://10.10.10.10, I have configured an URL Override, which will translate this to "cmd://PuTTY.exe -ssh {USERNAME}@10.10.10.10 -pw {PASSWORD}" --> Basically KeePass opens a SSH session with PuTTY from the command line. Super simple. I do the same with TeamViewer, but they actually support receiving the password in BASE64 format, adding more security to the process. KeePass of course can send passwords in encrypted formats.

So for a DIY integration with KeePass, you would just need to add a CLI set of commands to launch a session. If it can be launchend through the CLI, it can be launched in KeePass by simply hitting Ctrl+U.

There a number of Placeholders that KeePass can use, the list is quite comprehensive: https://keepass.info/help/base/placeholders.html

Let me know what you guys think. I've successfully integrated a number of programs doing this, RDP, TeamViewer, PuTTY and WinSCP to name a few.

**Comment by d18c7db** [comment]
> KeePass allows you to run commands from the URL field.

Actually this may be a good initial starting point and also pretty simple from a coding perspective, as described above, allow mRemoteNG to receive the authentication credentials and start a new session from the command line. If mRemoteNG  is already running then the new instance opens in a new tab of the already running mRemoteNG.

This way you don't need to give mRemoteNG your Keepass master password, since you start a new session from inside Keepass and mRemoteNG just receives only what it needs.

Also since we all know you can't keep everyone happy all the time, users have the option to either use the command line switches or not, no one has to do anything they don't want to. I suspect this would be a good compromise for those asking for "mRemoteNG Keepass integration"

**Comment by xercesbluebutterfly** [comment]
+1
